File: rc4.py
def parse_key_vector(key_value: str) -> list:
    if key_value.startswith("0x"):
        key_value = key_value[2:]  
    
    key = list()
    i = 0
    while i < len(key_value):
        key.append(
            int(key_value[i], 16) * 16
            + int(key_value[i+1], 16)
        )
        i += 2
        
    return key

def initialize_state_vector(size: int) -> list:
    state_vector = list()
    
    for i in range(size):
        state_vector.append(i)
        
    return state_vector

def swap(vector: list, source: int, dest: int) -> list:
    temp = vector[dest]
    vector[dest] = vector[source]
    vector[source] = temp
    
    return vector

def ksa(key_value: str, vectorlen: int = 256) -> list:
    key = parse_key_vector(key_value)
    state_vector = initialize_state_vector(vectorlen // 8)    
    
    keylen = len(key)
    svlen = len(state_vector)
    
    j = 0
    for i in range(vectorlen // 8):
        j = (j + state_vector[i] + key[i % keylen]) % svlen
        swap(state_vector, i, j)
    
    print(state_vector)
    return state_vector
    
def prga(input: str, state_vector: list) -> str:
    text = list()
    vec = list(reversed(input))
    i = 0
    j = 0
    svlen = len(state_vector)
    
    while len(vec):
        i = (i + 1) % svlen
        j = (j + state_vector[i]) % svlen
        swap(state_vector, i, j)
        
        k = state_vector[(state_vector[i] + state_vector[j]) % svlen]
        c = vec.pop()
        text.append(chr(ord(c) ^ k))
                
    return "".join(list(text))

def rc4(input: str, key_value: str, vectorlen: int = 256) -> str:
    state_vector = ksa(key_value, vectorlen)
    encrypted = prga(input, state_vector)
    return encrypted

if __name__ == "__main__":
    text = input("text to encrypt: ")
    key_value = input("key value (hex): ") .lower()
    vectorlen = int(input("vector length (bits): "))
    
    assert vectorlen % 8 == 0
    if not (text.startswith("\"") and text.endswith("\"")):
        # numerical value, not ascii text
        # todo this idc
        pass
        
    print(f"RC4(\"{text}\"): {rc4(text, key_value, vectorlen)}")

File: rc4.cc
#include <iostream>
#include <vector>
#include <stack>
#include <stdint.h>
#include <sstream>

template <typename T>
std::ostream &operator<<(std::ostream &os, const std::vector<T> &vec)
{
    os << "[ ";
    for (T elem : vec)
        os << (int)elem << " ";

    return os << "]";
}

class RC4
{
private:
    uint8_t m_numbits;
    std::vector<uint8_t> m_key;
    std::vector<uint64_t> m_statevector;

    mutable std::string m_text;
    mutable std::string m_cipher;

public:
    RC4(uint8_t numbits, std::string inkey) : m_numbits(numbits)
    {
        std::string key = inkey;

        if (inkey.substr(0, 2) == "0x")
            key = inkey.substr(2, inkey.size());

        if (key.size() % 2 != 0)
            inkey = "0" + key;

        for (int i = 0; i < key.size(); i += 2)
        {
            /* --------- LOOOOL ---------- */
            // cpp type of problem
            char a = key[i];
            char b = key[i + 1];
            std::string str = std::string({a, b});
            /* --------------------------- */

            uint8_t byte = _parse_hexstring(str);
            m_key.push_back(byte);
        }
    }

    void KSA()
    {
        _initialize_statevector();
        uint8_t keylen = m_key.size();
        uint8_t svlen = m_statevector.size();

        uint64_t j = 0;
        for (int i = 0; i < svlen; i++)
        {
            j = (j + m_statevector[i] + m_key[i % keylen]) % svlen;
            this->swap<uint64_t>(m_statevector, i, j);
        }
    }

    void PRGA(const std::string &text)
    {
        std::stack<char> st;
        for (int i = text.size() - 1; i >= 0; i--)
            st.push(text[i]);

        m_text = text;
        m_cipher.clear();

        uint8_t svlen = m_statevector.size();
        char c;
        uint64_t i = 0, j = 0, k;

        while (st.size())
        {
            i = (i + 1) % svlen;
            j = (j + m_statevector[i]) % svlen;
            this->swap(m_statevector, i, j);

            k = m_statevector[(m_statevector[i] + m_statevector[j]) % svlen];
            c = st.top();
            st.pop();

            m_cipher += c ^ (char)k;
        }
    }

    std::string encrypt(const std::string &text)
    {
        KSA();
        PRGA(text);

        return m_cipher;
    }

    friend std::ostream &operator<<(std::ostream &os, const RC4 &obj)
    {
        os << "Key vector: " << obj.m_key << std::endl;
        os << "State vector: " << obj.m_statevector << std::endl;
        os << "Text: " << obj.m_text << std::endl;
        os << "Cipher/encrypted: " << obj.m_cipher;

        return os;
    }

private:
    uint8_t _parse_hexstring(std::string str)
    {
        return (uint8_t)std::stol(str, nullptr, 16);
    }

    void _initialize_statevector()
    {
        for (int i = 0; i < m_numbits / 8; i++)
            m_statevector.push_back(i);
    }

    template <typename T>
    void swap(std::vector<T> &ctr, uint8_t src, uint8_t dest)
    {
        T temp = ctr[dest];
        ctr[dest] = ctr[src];
        ctr[src] = temp;
    }
};

int main(int argC, char **argV)
{
    std::string key = "", text = "";
    uint8_t bits = 0;

    for (int i = 1; i < argC; i++)
    {
        std::string str = std::string(argV[i]);

        if (str == "-k" || str == "--key")
            key = argV[i + 1];
        else if (str == "-b" || str == "--bits")
            bits = std::stoi(argV[i + 1], nullptr, 10);
        else if (str == "-t" || str == "--text")
            text = argV[i + 1];
    }

    if (text == "")
    {
        std::cout << "text to encrypt: ";
        std::getline(std::cin, text);
    }

    if (key == "")
    {
        std::cout << "key value (hex): ";
        std::getline(std::cin, key);
    }

    if (bits == 0)
    {
        std::cout << "vector length (bits): ";
        std::cin >> bits;
        std::cin.ignore(1); // newl
    }

    RC4 algo = RC4(bits, key);
    algo.encrypt(text);
    std::cout << algo << std::endl;
}